qemu_test: disable features that are not needed for tests (closure 641 -> 335.3M)

andir opened the pull request:
###### Motivation for this change

This is an attempt to change the NixOS test runner in a way that allows faster iteration on lower-level parts of the system by reducing the number of packages that contribute to the test closure size.

While `qemu_test` has always been a bit slimmer then one of its siblings it wasn't really "slim" in comparison. With this PR we reduce the closure size of `qemu_test` to about 335MB from previously > 640MB.

One of the changes this introduces (based on the discussion when this PR was first opened) is that test drivers will now come without any graphical support by default. There is a new attribute to the tests (`driverInteractive`) that comes with the full feature set of QEMU including the graphical output and sound.

Also, while running VM tests, we will use the QEMU Guest Agent from the corresponding QEMU package as otherwise we'd not have gained much.

I removed `systemd ` from `iputils` as that was only used to toggle the systemd unit generation (that we aren't using on systemd anyway). There are no references to systemd in the runtime closure. The required patch has been submitted upstream and is awaiting feedback.

###### Things done
- I executed a few tests nothing more fancy. What are good candidates to verify it doesn't break more fancy tests?
   - [X] tests.installer.simple.x86_64-linux
   - [X] tests.installer.zfsroot.x86_64-linux
   - [X] tests.vault.x86_64-linux
   - [X] tests.mpd.x86_64-linux
   - [X] tests.containers-ipv4.x86_64-linux

GrahamcOfBorg commented:
<!--REQUEST_ID=226ca480-1446-46e3-8def-feaa3b22fd5d-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49403&attempt_id=90a439d0-cc03-4d7d-a4f2-27aad172047a)

Attempted: qemu_test

<details><summary>Partial log (click to expand)</summary><p>

```
find: '/nix/store/zri6x2b20rjg6avdddsa8k8vrwsna802-qemu-host-cpu-only-for-vm-tests-3.0.0-ga': No such file or directory
find: '/nix/store/zri6x2b20rjg6avdddsa8k8vrwsna802-qemu-host-cpu-only-for-vm-tests-3.0.0-ga': No such file or directory
find: '/nix/store/zri6x2b20rjg6avdddsa8k8vrwsna802-qemu-host-cpu-only-for-vm-tests-3.0.0-ga': No such file or directory
find: '/nix/store/zri6x2b20rjg6avdddsa8k8vrwsna802-qemu-host-cpu-only-for-vm-tests-3.0.0-ga': No such file or directory
find: '/nix/store/zri6x2b20rjg6avdddsa8k8vrwsna802-qemu-host-cpu-only-for-vm-tests-3.0.0-ga': No such file or directory
find: '/nix/store/zri6x2b20rjg6avdddsa8k8vrwsna802-qemu-host-cpu-only-for-vm-tests-3.0.0-ga': No such file or directory
find: '/nix/store/zri6x2b20rjg6avdddsa8k8vrwsna802-qemu-host-cpu-only-for-vm-tests-3.0.0-ga': No such file or directory
find: '/nix/store/zri6x2b20rjg6avdddsa8k8vrwsna802-qemu-host-cpu-only-for-vm-tests-3.0.0-ga': No such file or directory
find: '/nix/store/zri6x2b20rjg6avdddsa8k8vrwsna802-qemu-host-cpu-only-for-vm-tests-3.0.0-ga': No such file or directory
/nix/store/f53dznvx2k7sy9r4a2q83yibhsdy49ar-qemu-host-cpu-only-for-vm-tests-3.0.0
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=226ca480-1446-46e3-8def-feaa3b22fd5d-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49403&attempt_id=75e602f4-2bb9-4e3a-add8-86056b483c6c)

Attempted: qemu_test

<details><summary>Partial log (click to expand)</summary><p>

```
patching script interpreter paths in /nix/store/ibz4sprmrmvs6r2plwgalsdnslapymvn-qemu-host-cpu-only-for-vm-tests-3.0.0
checking for references to /build in /nix/store/ibz4sprmrmvs6r2plwgalsdnslapymvn-qemu-host-cpu-only-for-vm-tests-3.0.0...
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
/nix/store/ibz4sprmrmvs6r2plwgalsdnslapymvn-qemu-host-cpu-only-for-vm-tests-3.0.0
```
</p></details>

xeji commented:
This removes the possibility to log in to testing machines interactively `nix-build some-test.nix -A driver` and then running `result/bin/nixos-test-driver` or `result/bin/nixos-run-vms`, which I think is a crucial feature for developing and debugging nixos tests. 

It also breaks `nixos-build-vms` which uses `qemu_test` by default because the resulting VMs are headless and non-interactive.

GrahamcOfBorg commented:
<!--REQUEST_ID=226ca480-1446-46e3-8def-feaa3b22fd5d-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.49403&attempt_id=3877d7b4-8058-4a2e-a320-e7e8035efa31)

Attempted: qemu_test

<details><summary>Partial log (click to expand)</summary><p>

```
patching script interpreter paths in /nix/store/bj6xlpiqcsf1ds5225fczhxyrriyqi47-qemu-host-cpu-only-for-vm-tests-3.0.0
checking for references to /build in /nix/store/bj6xlpiqcsf1ds5225fczhxyrriyqi47-qemu-host-cpu-only-for-vm-tests-3.0.0...
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
cannot find section .dynamic
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
/nix/store/bj6xlpiqcsf1ds5225fczhxyrriyqi47-qemu-host-cpu-only-for-vm-tests-3.0.0
```
</p></details>

andir commented:
@xeji true, I haven't thought about them. Do we have some intermediate path that we can go without pulling in all those dependencies for all the tests? When iterating on things like libc, systemd, … it is a bit annoying to rebuild gtk, pulseaudio, and a few other packages.

xeji commented:
> Do we have some intermediate path that we can go without pulling in all those dependencies for all the tests?

I'm not sure there is. Maybe it's better to have two variants `qemu_test_minimal` (the minimum needed to run the automated tests on Hydra) and `qemu_test` (for `nixos-build-vms` and developing/building nixos tests. But that would require some refactoring.

andir commented:
The difference is that there will not be a g GTK window popping up. You can still connect to localhost via a vnclient to get the same output.

andir commented:
After a bit of discussion with @xeji on IRC we think this might be fine for working on nixpkgs but we should implement something better for `nixos-build-vms` since in those cases the users most likely wants to see the VMs

flokli commented:
I'd really like to see this. Not having to rebuild all of GTK and its dependencies, to run a VM test after a low-level change (like systemd or glibc) is a huge timesaver.

tfc commented:
Oh, awesome. I see this PR for the first time right now. Why wasn't it merged earlier? Was the difficulty of testing if the change breaks any existing tests a/the problem?

flokli commented:
> Oh, awesome. I see this PR for the first time right now. Why wasn't it merged earlier? Was the difficulty of testing if the change breaks any existing tests a/the problem?

No, the problem was that people want to see the GTK window when interactively debugging (via `result/bin/nixos-run-vms`, and maybe also via the test driver), but there's currently no way to distinguish.

andir committed:
qemu: strip down the features for the test runner

This allows much faster VM-test based systemd testing as the closure of
qemu suddenly shrinks to reasonable sizes again.

andir committed:
nixos/tests: make the `driver` attribute use a rich qemu

Since we previously stripped down the features of `qemu_test` some of
the features users are used to while running tests through the (impure)
driver didn't work anymore. Most notably we lost support for graphical
output and audio. With this change the `driver` attribute uses are more
feature complete version of QEmu compared to the one used in the pure
Nix builds.

This gives us the best of both worlds. Users are able to see the
graphical windows of VMs while CI and regular nix builds do not have to
download all the (unnecessary) dependencies.

andir committed:
nixos/tests: expose both the interactive and non-interactive driver

For a lot of the work the non-interactive drivers are enough and it is
probably a good idea to keep it accessible for debugging without
touching the Nix expression.

andir commented:
I've updated this PR to my latest work in this area.

The current `qemu_test` size is now 340.9M down from 641.M before this PR. I'll look into reducing the size of qemu closure further now.

andir committed:
qemu: make ncurses optional for the test runner

This shaves another 3MB off the closure size of QEMU.

andir committed:
nixos/qemu-guest-agent: make the QEMU guest agent package configurable

andir committed:
nixos/tests: make sure we use the qemu_test package to provide the Guest Agent

This reduces the closure size for the minimal test by a lot since we no
longer have to build the regular QEMU for even the simplest test.

andir committed:
iptuils: Remove systemd from dependencies

In our case systemd is only used to figure out if the unit files should
be generated.

andir committed:
iputils: apply proposed upstream patch to get rid of systemd dependency

This allows us to get rid of the hack and the systemd dependency and
thus reduces the rebuild closure whenever systemd changes.

andir committed:
libndctl: clarify that it needs udev not systemd

While (currently) it is the same package it carries more information if
we explicitly state that we want udev.

andir commented:
I updated the initial description:

###### Motivation for this change

This is an attempt to change the NixOS test runner in a way that allows faster iteration on lower-level parts of the system by reducing the number of packages that contribute to the test closure size.

While `qemu_test` has always been a bit slimmer then one of its siblings it wasn't really "slim" in comparison. With this PR we reduce the closure size of `qemu_test` to about 335MB from previously > 640MB.

One of the changes this introduces (based on the discussion when this PR was first opened) is that test drivers will now come without any graphical support by default. There is a new attribute to the tests (`driverInteractive`) that comes with the full feature set of QEMU including the graphical output and sound.

Also, while running VM tests, we will use the QEMU Guest Agent from the corresponding QEMU package as otherwise we'd not have gained much.

I removed `systemd ` from `iputils` as that was only used to toggle the systemd unit generation (that we aren't using on systemd anyway). There are no references to systemd in the runtime closure. The required patch has been submitted upstream and is awaiting feedback.

###### Things done
- I executed a few tests nothing more fancy. What are good candidates to verify it doesn't break more fancy tests?
   - [X] tests.installer.simple.x86_64-linux
   - [X] tests.installer.zfsroot.x86_64-linux
   - [X] tests.vault.x86_64-linux
   - [X] tests.mpd.x86_64-linux
   - [X] tests.containers-ipv4.x86_64-linux

flokli commented:
@andir I also did successfully run `nixosTests.cage` (which does screenshots and OCR), `nixosTests.bittorrent` (which does some more fancy networking stuff with multiple switches), and `nixosTests.signal-desktop` (GUI, but X).

IMHO, this should be good enough (tm) to merge. Do you want to squash 54e6cfc1efd8e563371fee815389e51554998eef and 0a55c5dada75480342ee9b58932d612438793302 together?

andir commented:
I would like to keep those two commits on iputils as is. It gives a bit more history on how this came to be. It might also be useful to revert the commit that applies the commit if upstream doesn't accept it (then we already have an alternative implementation).

I did rebuild all the release tests and only had a +/-1 deviation from the regular amount of test failures.

rnhmjoj commented:
It seems this change has been applied to `nixos-rebuild build-vm` too. I hope it's not intended because I found invaluable to simply start the SDL screen without using a remote viewer.

andir commented:
On 01:54 21.10.20, Michele Guerini Rocco wrote:
> It seems this change has been applied to `nixos-rebuild build-vm` too.
> I hope it's not intended because I found invaluable to simply start
> the SDL screen without using a remote viewer.

Oh that is by accident. I'll have a look on how that can be fixed.

xfix commented:
@ofborg test installer.simple
